contracts/adapters/Distribute.sol

```diff
@@ -0,0 +1,280 @@
+     * @param proposalId The distribution proposal id.
+     */
+    function processProposal(DaoRegistry dao, bytes32 proposalId)
+        external
+        override
+    {
+        dao.processProposal(proposalId);
+
+        // Checks if the proposal exists or is not in progress yet.
+        Distribution storage distribution =
+            distributions[address(dao)][proposalId];
+        require(
+            distribution.status == DistributionStatus.NOT_STARTED,
+            "proposal already completed or in progress"
+        );
+
+        // Checks if there is an ongoing proposal, only one proposal can be executed at time.
+        bytes32 ongoingProposalId = ongoingDistributions[address(dao)];
+        require(
+            ongoingProposalId == bytes32(0) ||
+                distributions[address(dao)][ongoingProposalId].status !=
+                DistributionStatus.IN_PROGRESS,
+            "another proposal already in progress"
+        );
+
+        // Checks if the proposal has passed.
+        IVoting votingContract = IVoting(dao.getAdapterAddress(VOTING));
+        require(
+            votingContract.voteResult(dao, proposalId) ==
+                IVoting.VotingState.PASS,
+            "proposal did not pass"
+        );
+
+        distribution.status = DistributionStatus.IN_PROGRESS;
+        distribution.blockNumber = block.number;
+        ongoingDistributions[address(dao)] = proposalId;
+    }
+
+    /**
+     * @notice Transfers the funds from the Guild account to the member's internal accounts.
+     * @notice The amount of funds is caculated using the historical number of shares of each member.
+     * @dev A distribution proposal must be in progress.
+     * @dev Only proposals that have passed the voting can be completed.
+     * @dev Only active members can receive funds.
+     * @param dao The dao address.
+     * @param toIndex The index to control the cached for-loop.
+     */
+    function distribute(DaoRegistry dao, uint256 toIndex) external override {
+        // Checks if the proposal does not exist or is not completed yet
+        bytes32 ongoingProposalId = ongoingDistributions[address(dao)];
+        Distribution storage distribution =
+            distributions[address(dao)][ongoingProposalId];
+        uint256 blockNumber = distribution.blockNumber;
+        require(
+            distribution.status == DistributionStatus.IN_PROGRESS,
+            "distribution completed or does not exist"
+        );
+
+        // Check if the given index was already processed
+        uint256 currentIndex = distribution.currentIndex;
+        require(currentIndex <= toIndex, "toIndex too low");
+
+        address token = distribution.token;
+        uint256 amount = distribution.amount;
+
+        // Get the total number of shares when the proposal was processed.
+        BankExtension bank = BankExtension(dao.getExtensionAddress(BANK));
+        uint256 totalShares = bank.getPriorAmount(TOTAL, SHARES, blockNumber);
```

Comment:
let's move totalShares to the "else" part where it is used so we don't do this call if not needed